crates/algo_models/src/lib.rs

```diff
@@ -457,112 +479,196 @@ impl AlgorandMsgpack for SignedTransaction {
+
+        let encoded_struct = tx_struct.encode().unwrap();
+        let decoded_struct = AssetTransferTransactionFields::decode(&encoded_struct).unwrap();
+        assert_eq!(decoded_struct, tx_struct);
+
+        let tx_enum = Transaction::AssetTransfer(tx_struct.clone());
+        let encoded_enum = tx_enum.encode().unwrap();
+        let decoded_enum = Transaction::decode(&encoded_enum).unwrap();
+        assert_eq!(decoded_enum, tx_enum);
+
+        let signed_tx = SignedTransaction {
+            transaction: tx_enum.clone(),
+            signature: [0; 64],
+        };
+        let encoded_stx = signed_tx.encode().unwrap();
+        let decoded_stx = SignedTransaction::decode(&encoded_stx).unwrap();
+        assert_eq!(decoded_stx, signed_tx);
+        assert_eq!(decoded_stx.transaction, tx_enum);
+
+        let raw_encoding = tx_struct.encode_raw().unwrap();
+        assert_eq!(encoded_struct[0], b'T');
+        assert_eq!(encoded_struct[1], b'X');
+        assert_eq!(encoded_struct.len(), raw_encoding.len() + 2);
+        assert_eq!(encoded_struct[2..], raw_encoding);
+        assert_eq!(encoded_struct.len(), 117);
+    }
 
-    let addr_from_str = Address::from_string(&addr.address()).unwrap();
-    assert_eq!(addr, addr_from_str);
+    #[test]
+    fn test_address() {
+        let addr = Address::from_pubkey(&[0; 32]);
+        assert_eq!(
+            addr.address(),
+            "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ"
+        );
+
+        let addr_from_str = Address::from_string(&addr.address()).unwrap();
+        assert_eq!(addr, addr_from_str);
+    }
+
+    #[test]
+    fn test_pay_transaction_raw_id() {
+        let expected_tx_id = [
+            35, 93, 0, 170, 96, 221, 1, 74, 119, 147, 131, 116, 7, 31, 225, 40, 215, 47, 44, 120,
+            128, 245, 41, 65, 116, 255, 147, 64, 90, 80, 147, 223,
+        ];
+
+        let payment = example_pay_transaction();
+        let unsigned_tx = Transaction::Payment(payment.clone());
+        let signed_tx = SignedTransaction {
+            transaction: unsigned_tx.clone(),
+            signature: [0; 64],
+        };
+
+        assert_eq!(payment.raw_id().unwrap(), expected_tx_id);
+        assert_eq!(unsigned_tx.raw_id().unwrap(), expected_tx_id);
+        assert_eq!(signed_tx.raw_id().unwrap(), expected_tx_id);
+    }
+
+    #[test]
+    fn test_pay_transaction_id() {
+        let expected_tx_id = "ENOQBKTA3UAUU54TQN2AOH7BFDLS6LDYQD2SSQLU76JUAWSQSPPQ";
+
+        let payment = example_pay_transaction();
+        let unsigned_tx = Transaction::Payment(payment.clone());
+        let signed_tx = SignedTransaction {
+            transaction: unsigned_tx.clone(),
+            signature: [0; 64],
+        };
+
+        assert_eq!(payment.id().unwrap(), expected_tx_id);
+        assert_eq!(unsigned_tx.id().unwrap(), expected_tx_id);
+        assert_eq!(signed_tx.id().unwrap(), expected_tx_id);
+    }
+
+    fn example_pay_transaction() -> PayTransactionFields {
```

Comment:
Curious if there is some conventional way in rust to represent test data?
I was thinking about creating a test data module, however figured i'd push something simple for now.

Reply:
Good question. Not sure what the idiomatic way to do this is. Also worth noting there is a test data JSON file: https://github.com/algorandfoundation/algokit-core/blob/main/crates/algokit_transact_ffi/test_data.json

This file is generated via this script: https://github.com/algorandfoundation/algokit-core/blob/main/scripts/generate_test_data/index.ts

Maybe it would make sense to use that same data in the rust unit tests? 

Regardless, I think what you have now is fine and we can refine test strategy in a future PR

Reply:
Yeah I did a bit of research too, and not much came up.
Potentially we could explore using an Object Mother pattern, which could also expand to include some test data builders.

It probably does make sense to share the test data.

For now I'll leave as is.